x-pack/auditbeat/processors/sessionmd/provider/modernprovider/modernprovider_linux.go

```diff
@@ -0,0 +1,523 @@
+		if now.Sub(since) > resetDuration {
+			since = now
+			combinedWait = 0 * time.Millisecond
+		}
+	}
+
+	start := now
+	nextWait := 5 * time.Millisecond
+	for {
+		waited := time.Since(start)
+		if _, found := p.lookupLocked(pid); found {
+			p.logger.Debugf("got process that was missing after %v", waited)
+			combinedWait = combinedWait + waited
+			return nil
+		}
+		if waited >= maxWaitLimit {
+			e := fmt.Errorf("process %v was not seen after %v", pid, waited)
+			p.logger.Warnf("%w", e)
+			combinedWait = combinedWait + waited
+			return e
+		}
+		time.Sleep(nextWait)
+		if nextWait*2+waited > maxWaitLimit {
+			nextWait = maxWaitLimit - waited
+		} else {
+			nextWait = nextWait * 2
+		}
+	}
+}
+
+func (p prvdr) GetProcess(pid uint32) (*types.Process, error) {
+	proc, found := p.lookupLocked(pid)
+	if !found {
+		return nil, fmt.Errorf("PID %d not found in cache", pid)
+	}
+
+	interactive := interactiveFromTTY(types.TTYDev{
+		Major: proc.Proc.TtyMajor,
+		Minor: proc.Proc.TtyMinor,
+	})
+
+	start := time.Unix(0, int64(proc.Proc.TimeBoot))
+
+	ret := types.Process{
+		PID:              proc.Pid,
+		Start:            &start,
+		Name:             basename(proc.Filename),
+		Executable:       proc.Filename,
+		Args:             proc.Cmdline,
+		WorkingDirectory: proc.Cwd,
+		Interactive:      &interactive,
+	}
+
+	euid := proc.Proc.Euid
+	egid := proc.Proc.Egid
+	ret.User.ID = strconv.FormatUint(uint64(euid), 10)
+	username, ok := getUserName(ret.User.ID)
+	if ok {
+		ret.User.Name = username
+	}
+	ret.Group.ID = strconv.FormatUint(uint64(egid), 10)
+	groupname, ok := getGroupName(ret.Group.ID)
+	if ok {
+		ret.Group.Name = groupname
+	}
+	ret.TTY.CharDevice.Major = uint16(proc.Proc.TtyMajor)
+	ret.TTY.CharDevice.Minor = uint16(proc.Proc.TtyMinor)
+	if proc.Exit.Valid {
+		end := time.Unix(0, int64(proc.Exit.ExitTimeProcess))
+		ret.ExitCode = proc.Exit.ExitCode
+		ret.End = &end
+	}
+	ret.EntityID = calculateEntityIDv1(pid, *ret.Start)
+
+	p.fillParent(&ret, proc.Proc.Ppid)
+	p.fillGroupLeader(&ret, proc.Proc.Pgid)
+	p.fillSessionLeader(&ret, proc.Proc.Sid)
+	p.fillEntryLeader(&ret, proc.Proc.EntryLeader)
+	setEntityID(&ret)
+	setSameAsProcess(&ret)
+	return &ret, nil
+}
+
+func (p prvdr) lookupLocked(pid uint32) (quark.Process, bool) {
+	p.qqMtx.Lock()
+	defer p.qqMtx.Unlock()
+
+	return p.qq.Lookup(int(pid))
+}
+
+func (p prvdr) fillParent(process *types.Process, ppid uint32) {
+	proc, found := p.lookupLocked(ppid)
+	if !found {
+		return
+	}
+
+	start := time.Unix(0, int64(proc.Proc.TimeBoot))
+	interactive := interactiveFromTTY(types.TTYDev{
+		Major: proc.Proc.TtyMajor,
+		Minor: proc.Proc.TtyMinor,
+	})
+	euid := proc.Proc.Euid
+	egid := proc.Proc.Egid
+	process.Parent.PID = proc.Pid
+	process.Parent.Start = &start
+	process.Parent.Name = basename(proc.Filename)
+	process.Parent.Executable = proc.Filename
+	process.Parent.Args = proc.Cmdline
+	process.Parent.WorkingDirectory = proc.Cwd
```

Comment:
`Filename`, `Cmdline` and `Cwd` can be empty strings, just writing it here to make sure it is accounted for.